### PR TITLE
feat: sequence aware mempool prototype

### PR DESCRIPTION
**Pull request by cmwaters**
A priority mempool based on gas price has always been incompatible with the replay protection mechanism that relies on monotonically increasing sequences. This becomes increasingly apparent when the same signer is submitting multiple blobs per block.

I have written an ADR proposing a simple-ish solution which bundles pending transactions per signer. I also have written a working prototype (some of it vibe-coded) which solves the issue with the tx client test here: https://github.com/celestiaorg/celestia-app/issues/5655

### CODE REVIEW

File: mempool/cat/store.go [hunk 1]
```diff
@@ -229,7 +311,96 @@ func (s *store) getOrderedTxs() []*wrappedTx {
...
+				} else {
+					set.aggregatedPriority = 0
+					set.firstTimestamp = time.Time{}
+				}
+			} else {
+				set.aggregatedPriority = 0
+				set.firstTimestamp = time.Time{}
+			}
+			return true
+		}
```
Comment (evan-forbes) [inline mempool/cat/store.go]: [flup]
can we clean this up or try to make it more readable?

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+## Decision
+
+Group transactions by signer into signer-bundled sets ("tx sets") and order based on average weighted priority per set. Within tx set order transactions in ascending sequence order. Block building and recheckTx iterates through transactions in those order. This will preserve sequence ordering per signer while maximising tx fees per block.
+
+Key aspects:
+
+- Signer-bundled sets: A `txSet` holds transactions from a single signer. The set tracks:
+    - `txs`: maintained in ascending sequence order; ties broken by earlier arrival timestamp.
+    - `aggregatedPriority`: gas-weighted average of member transaction priorities, floored to `int64`.
```
Comment (rach-id) [inline docs/celestia-architecture/adr-011-sequence-aware-cat.md:25]: since the priority is the average, couldn't someone post very cheap transactions at the beginning, and have the 100th transaction with crazy high fee forcing you to include the first ones, but when you arrive to the last, the block space is full 😆 

What do you think of switching to median?

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md
Comment (rach-id) [inline docs/celestia-architecture/adr-011-sequence-aware-cat.md:1]: can we run some benchmarks while we're at it to see if this is introducing any delays in processing transactions?

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    - `bytes`, `firstTimestamp`, and running accumulators `totalGasWanted` and `weightedPrioritySum`.
+- Global ordering of sets: `orderedTxSets` is kept sorted by `(aggregatedPriority desc, firstTimestamp asc)`. This preserves FIFO between sets of equal aggregated priority.
+- Within-set ordering: Transactions are iterated in ascending sequence, guaranteeing that a later sequence from the same signer is never submitted before an earlier one.
+- Eviction policy when full: Compute the new set aggregated priority if the incoming transaction were added (`aggregatedPriorityAfterAdd`). Identify victim sets with aggregated priority strictly below the incoming set. Evict transactions from victim sets starting from the end (highest sequence numbers first) until enough bytes are freed. If insufficient bytes can be freed, reject the incoming transaction.
+- TTL purge: Expiration by height/time removes transactions from sets and reorders or removes empty sets accordingly.
```
Comment (rach-id) [inline docs/celestia-architecture/adr-011-sequence-aware-cat.md:30]: when purging after TTL, we could delete the earliest nonce transaction while the ones after it are not removed, right?

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        - `bytes int64`
+        - `firstTimestamp time.Time`
+        - `totalGasWanted int64`, `weightedPrioritySum int64`
+    - `store`:
+        - `setsBySigner map[string]*txSet`
+        - `orderedTxSets []*txSet` (sorted by weighted average priority)
+        - transaction index, byte accounting, reservation tracking
+
+- Aggregation:
+    - On insert: update aggregated prioritt `totalGasWanted += gasWanted`, `weightedPrioritySum += priority * gasWanted`, then `aggregatedPriority = weightedPrioritySum / totalGasWanted` (integer division). Aggregated priority is relative to the amount of gasWanted per transaction. Transactions with more gasWanted will have larger weighting. Alternatively we could have used transaction size.
```
Comment (rach-id) [inline docs/celestia-architecture/adr-011-sequence-aware-cat.md]:
```suggestion
    - On insert: update aggregated priority `totalGasWanted += gasWanted`, `weightedPrioritySum += priority * gasWanted`, then `aggregatedPriority = weightedPrioritySum / totalGasWanted` (integer division). Aggregated priority is relative to the amount of gasWanted per transaction. Transactions with more gasWanted will have larger weighting. Alternatively we could have used transaction size.
```

File: mempool/cat/pool.go [hunk 5]
```diff
@@ -334,10 +334,11 @@ func (txmp *TxPool) TryAddNewTx(tx *types.CachedTx, key types.TxKey, txInfo memp
 		txmp.metrics.FailedTxs.Add(1)
 		return rsp, nil
 	}
+	// removed noisy print used during debugging
```
Comment (rach-id) [inline mempool/cat/pool.go]:
```suggestion
```

File: mempool/cat/pool.go [hunk 6]
```diff
@@ -461,6 +462,7 @@ func (txmp *TxPool) Update(
 	newPreFn mempool.PreCheckFunc,
 	newPostFn mempool.PostCheckFunc,
 ) error {
+	// removed noisy print used during debugging
```
Comment (rach-id) [inline mempool/cat/pool.go]:
```suggestion
```

File: mempool/cat/pool.go [hunk 7]
```diff
@@ -525,44 +527,46 @@ func (txmp *TxPool) addNewTransaction(wtx *wrappedTx) error {
...
+	outer:
+		for _, set := range victimSets {
+			// iterate in reverse order removing the higher sequence numbers first
+			for i := len(set.txs) - 1; i >= 0; i-- {
+				tx := set.txs[i]
+				txmp.evictTx(tx)
+				availableBytes += tx.size()
+				if availableBytes >= wtx.size() {
+					break outer
+				}
```
Comment (rach-id) [inline mempool/cat/pool.go]: can this be reorganized to avoid having a tag? it becomes harder to understand the logic

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md (hunk 2 above)
Comment (tzdybal) [reply docs/celestia-architecture/adr-011-sequence-aware-cat.md:25]: My initial intuition was to use "next tx gas" as `txSet` priority. So we can always extract the most profitable transactions from mempool and maintain ordering inside sets.

But `Increasing fees under congestion` section clarifies this for me.

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+- TTL purge:
+    - TTL will remove an entire transaction set (i.e. everything from a specific signer) since all following transactions would also become invalidated
+
+### Invariants and Guarantees
+
+- Within a signer, transactions are never emitted out of ascending sequence order.
+- Across signers, ordering is by set weighted average priority (desc), with earlier `firstTimestamp` breaking ties (FIFO between equal-priority sets).
+- Weighted average priority updates are consistent with gas-weighted means as transactions are added/removed.
+- Eviction never removes lower sequence numbers in a set before removing higher ones, preserving the ability to submit remaining sequences correctly.
```
Comment (tzdybal) [inline docs/celestia-architecture/adr-011-sequence-aware-cat.md:67]: nit: This is very important - I would mention this even in context section.

**Review by tzdybal — COMMENTED** [review]
I would emphasize two facts in context section:
 -  > Eviction never removes lower sequence numbers in a set before removing higher ones, preserving the ability to submit remaining sequences correctly.
 -  > Rather than replace by fee, users can improve the likeness of their transaction landing by increasing the gas price of later transactions, thus raising the aggregated priority. Given sequence awareness, a replace by fee protocol may easily be supported in the future.

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+## Context
+
+The CAT mempool previously ordered transactions strictly based on priority (fee / gas). This is incongruent with the replay protection mechanism which depends on the order of transactions by a signer having a monotonically increasing sequence number. This becomes more apparent with multiple transactions by the same signer in the same block.
+
+We need a mempool ordering algorithm that respects per-signer sequencing constraints while still maximising revenue through choosing the highest paid tranasctions (per byte or gas used).
+
+## Decision
+
+Group transactions by signer into signer-bundled sets ("tx sets") and order based on average weighted priority per set. Within tx set order transactions in ascending sequence order. Block building and recheckTx iterates through transactions in those order. This will preserve sequence ordering per signer while maximising tx fees per block.
```
Comment (rootulp) [inline docs/celestia-architecture/adr-011-sequence-aware-cat.md:19]: [question] Is it possible for an attacker to submit two txs:
1. tx 1 with sequence 1 and a low priority
2. tx 2 with sequence 2 and a high priority

such that the second tx increases the average weighted priority of the tx set such that tx 1 gets included on-chain, however once tx 1 gets included on-chain it invalidates tx 2 so tx 2 can no longer get on chain. The end result is that the signer got tx 1 included on-chain ahead of competing txs that would have paid more in fees?

Update: oops disregard, I see Rachid raised the same concern.

**Review by rootulp — COMMENTED** [review]
One idea to avoid the attack Rachid raised is instead of calculating the average weighted priority, preserve the logic to capture tx sets but only inspect the lowest sequence in each tx set when evaluating which txs to include in a block. Then choose txs to include based on priority.

In other words, I'm proposing a less sophisticated mempool that does not calculate average weighted priority but does respect the signer's need to get lower sequence txs included before higher sequence txs.

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md (hunk 9 above)
Comment (adlerjohn) [inline docs/celestia-architecture/adr-011-sequence-aware-cat.md:19]: Rather than an attack, is this an issue in the honest but congested case?

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md
Comment (cmwaters) [reply docs/celestia-architecture/adr-011-sequence-aware-cat.md:1]: Do we have existing benchmarks that would be good to understand any changes in the performance (or do we need to write new ones)

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md (hunk 2 above)
Comment (cmwaters) [reply docs/celestia-architecture/adr-011-sequence-aware-cat.md:25]: > couldn't someone post very cheap transactions at the beginning, and have the 100th transaction with crazy high fee forcing you to include the first ones, but when you arrive to the last, the block space is full 😆

The user would presumably still pay for the 100th transaction because it would just be included in the following block unless you can find a way to invalidate it.

A counter proposal to your median idea (which I think could also be effective):

Evaluate the priority based on the first n bytes worth of transactions (where n is less than the size of a block) rather than the aggregated priority of all the transactions

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md (hunk 9 above)
Comment (cmwaters) [reply docs/celestia-architecture/adr-011-sequence-aware-cat.md:19]: Responded to it in rachid's comment

Comment (cmwaters) [reply docs/celestia-architecture/adr-011-sequence-aware-cat.md:19]: Yes we need this mechanism in the case of congestion or more broadly whenever we expect priorities to be different

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md
Comment (rach-id) [reply docs/celestia-architecture/adr-011-sequence-aware-cat.md:1]: I don't think we have anything on main. I have personal ones I use for my benchmarks, but they're unmergeable 😄

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md (hunk 3 above)
Comment (cmwaters) [reply docs/celestia-architecture/adr-011-sequence-aware-cat.md:30]: No we'd remove an entire set all together - the reasoning is that all subsequent transactions would have invalid nonces and be rejected by recheck tx anyway

**Comment by cmwaters** [comment]
> One idea to avoid the attack Rachid raised is instead of calculating the average weighted priority, preserve the logic to capture tx sets but only inspect the lowest sequence in each tx set when evaluating which txs to include in a block. Then choose txs to include based on priority.

Can you perhaps pseudocode the design? So could only the first transaction with the lowest sequence make it into a block or would you be able to do multiple?

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md (hunk 2 above)
Comment (evan-forbes) [reply docs/celestia-architecture/adr-011-sequence-aware-cat.md:25]: > Evaluate the priority based on the first n bytes worth of transactions (where n is less than the size of a block) 

not sure I understand this. how does the first block worth of transaction determine the highest priority txs?

why not just evict an entire set if the one with the lowest nonce doesn't cut it?

**Review by evan-forbes — APPROVED** [review]
I'm happy to fuzz / debug / refactor this in app after. Also happy to flup the adr as well.

I'm in favor of shipping this simplest possible version of this **ASAP**

we need to fix tx submission

File: mempool/cat/store.go (hunk 1 above)
Comment (ninabarbakadze) [reply mempool/cat/store.go]: I already cleaned up everything I don't know why you are seeing the outdated version. The only reason why I had it in draft was the state of tests. There is barely any proper unit test coverage which I wanted to take time and write because we had memory leaks in the store before that was not caught by tests but I unfortunately got busy with the sovereign issue and tx client bug in evictions.

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md (hunk 2 above)
Comment (cmwaters) [reply docs/celestia-architecture/adr-011-sequence-aware-cat.md:25]: I'm confused. This isn't about eviction. It's about deciding the transactions to include in a block.

File: docs/celestia-architecture/adr-011-sequence-aware-cat.md
Comment (cmwaters) [reply docs/celestia-architecture/adr-011-sequence-aware-cat.md:1]: I'll see if I can find some or write them up myself

**Comment by ninabarbakadze** [comment]
> I'm happy to fuzz / debug / refactor this in app after. Also happy to flup the adr as well.
> 
> I'm in favor of shipping this simplest possible version of this **ASAP**
> 
> we need to fix tx submission

We made some changes mb take another look and we can start manually testing @evan-forbes. I opened a few FLUPs so we can track the changes we need in order to make the prototype fully production ready:
- https://github.com/celestiaorg/celestia-core/issues/2481
- https://github.com/celestiaorg/celestia-core/issues/2482

File: mempool/cat/store.go [hunk 10]
```diff
@@ -179,47 +245,49 @@ func (s *store) reset() {
...
+func (s *store) processOrderedTxSets(fn func(txSets []*txSet)) {
 	s.mtx.RLock()
 	defer s.mtx.RUnlock()
-	for _, tx := range s.orderedTxs {
-		if !fn(tx) {
-			break
-		}
-	}
+
+	fn(s.orderedTxSets)
```
Comment (evan-forbes) [inline mempool/cat/store.go:290]: not blocking **opinion**

generally I think it makes more sense to just not abstract this

imo a for look or mutex call is acceptable amounts of boilerplate

I know there are existing methods that follow this pattern, but it seems off that we are defining functions inside a method when we could just write a new method. again opinion and not blocking.

File: mempool/cat/tx.go [hunk 11]
```diff
@@ -35,3 +38,152 @@ func (w *wrappedTx) size() int64 { return int64(len(w.tx.Tx)) }
...
+		set.firstHeight = wtx.height
+		set.aggregatedPriority = set.weightedPrioritySum / set.totalGasWanted
+		return
+	}
+	idx := sort.Search(len(set.txs), func(i int) bool {
+		if set.txs[i].sequence == wtx.sequence {
+			return wtx.timestamp.Before(set.txs[i].timestamp)
+		}
+		return set.txs[i].sequence > wtx.sequence
+	})
```
Comment (evan-forbes) [inline mempool/cat/tx.go:87]: [not blocking but likely need a flup if not fixed here]

if we're going to cover the edge case of the sequence being the same, then shouldn't we being replacing  and not comparing timestamp?

it seemed like we went to a decent amount of trouble to compare timestamps in places/tests, so maybe we want to preserve that for some other reason?

File: mempool/cat/store_test.go [hunk 12]
```diff
@@ -53,81 +54,226 @@ func TestStoreOrdering(t *testing.T) {
...
+		require.Equal(t, 9, len(remainingTxs))           // One less transaction
+		require.NotContains(t, remainingTxs, txToRemove) // Removed tx not in list
+
+		// Verify sequences are still in order
+		for i := 0; i < 5; i++ {
+			require.Equal(t, uint64(i+1), remainingTxs[i].sequence)
+		}
+		for i := 5; i < 9; i++ {
+			require.Equal(t, uint64(i+2), remainingTxs[i].sequence)
+		}
```
Comment (evan-forbes) [inline mempool/cat/store_test.go:232]: [can be flupped]

I think I understand what this test is checking, but I have questions about this.

why two for loops? why are we adding 1 the first one and 2 in the second one? is that skipping a tx? if so why?

its a very minor worry, but since (afaiu) we do not expect GetOrderedTxs to always preserve sequence order, for example when we have different signers, do you think we could have a more robust test by not requiring that to be true and instead checking only sequence oder via txsets?

File: mempool/cat/store_test.go [hunk 13]
```diff
@@ -53,81 +54,226 @@ func TestStoreOrdering(t *testing.T) {
...
+		require.Equal(t, txs, orderedTxs)
+		require.Equal(t, 1, len(store.setsBySigner)) // One signer set
+		require.Equal(t, 10, len(store.setsBySigner[string([]byte("signer1"))].txs))
+
+		// Test removing middle transaction (sequence 6)
+		initialSize, initialBytes, initialSets := captureStoreState(store)
+		txToRemove := txs[5] // sequence 6
+
+		removed := store.remove(txToRemove.key())
+		require.True(t, removed)
```
Comment (evan-forbes) [inline mempool/cat/store_test.go:214]: what is the thing we are trying to test here? ofc, I see the testname "remove() non-empty signer set", but it doesn't seem like we're removing a txset, it seems like we're removing a specific tx that is already part of a larger set

 when would we remove the middle tx but not the txs above it?

File: mempool/cat/store_test.go [hunk 14]
```diff
@@ -53,81 +54,226 @@ func TestStoreOrdering(t *testing.T) {
...
+
+		// Verify signer set still exists but has one less transaction
+		signerSet := store.setsBySigner[string([]byte("signer1"))]
+		require.NotNil(t, signerSet)
+		require.Equal(t, 9, len(signerSet.txs))
```
Comment (evan-forbes) [inline mempool/cat/store_test.go:219]: [for my own understanding]

this doesn't match my current mental model for what we want to achieve with a nonce aware mempool. What's the usecase for removing a single tx from a set? why not remove all the txs above it as well since they cannot be included?

File: mempool/cat/store_test.go [hunk 15]
```diff
@@ -300,3 +473,223 @@ func TestStoreGetOrderedTxs(t *testing.T) {
...
+		for seq := 1; seq <= txsPerSigner; seq++ {
+			// Priority: start high for signer0, lower for signer1, etc.
+			priority := rand.Int63n(100)       // pick a random priority
+			gasWanted := rand.Int63n(1000) + 1 // pick a random gas wanted (not used here, but could be)
+			tx := types.Tx(fmt.Sprintf("tx_signer%d_seq%d", s, seq))
+			cachedTx := &types.CachedTx{Tx: tx}
+			wtx := newWrappedTx(cachedTx, 1, gasWanted, priority, signers[s], uint64(seq))
+			store.set(wtx)
+		}
+	}
```
Comment (evan-forbes) [inline mempool/cat/store_test.go:499]: [not blocking]

this is good! we might be able to improve it by testing for the exact properties we want instead of random. for example we could add txs with reverse priorities. 

its not a strong opinion, but imo if we're trying to test a set of specific properties for get order, it might benefit from a table driven test. We are effectively doing that throughout these tests! however the cases seems scattered. I could be wrong and just need to sleep on it. This is just my impression on the first read

for example in a comment above we're also checking for priority, which might mean we could organize the tests better. again, could be wrong here 🙂

File: mempool/cat/tx.go [hunk 16]
```diff
@@ -35,3 +38,152 @@ func (w *wrappedTx) size() int64 { return int64(len(w.tx.Tx)) }
...
+			set.totalGasWanted -= wtx.gasWanted
+			set.weightedPrioritySum -= wtx.priority * wtx.gasWanted
+			// Remove the tx from the set
+			set.txs = append(set.txs[:i], set.txs[i+1:]...)
+
+			// If the set is empty, or the total gas wanted is zero
+			// set the aggregated priority to zero
+			if len(set.txs) <= 0 || set.totalGasWanted <= 0 {
+				set.aggregatedPriority = 0
+				set.firstTimestamp = time.Time{}
```
Comment (evan-forbes) [inline mempool/cat/tx.go:123]: [not blocking]

if the txset is empty, why not remove it?

File: mempool/cat/store.go [hunk 17]
```diff
@@ -63,10 +93,40 @@ func (s *store) remove(txKey types.TxKey) bool {
...
 	}
 	s.bytes -= tx.size()
-	if err := s.deleteOrderedTx(tx); err != nil {
+	delete(s.txs, txKey)
+
+	// Update the signer's set
+	signerKey := string(tx.sender)
+	if signerKey == "" {
+		// Empty-signer sets are single-tx sets: remove the entire set containing tx
+		for i, set := range s.orderedTxSets {
```
Comment (evan-forbes) [inline mempool/cat/store.go:102]: do we have a test case that guarantees this?

edit: by this I mean "// Empty-signer sets are single-tx sets: remove the entire set containing tx"

do we have a test to guarantee that empty signer's are single tx sets

**Review by evan-forbes — APPROVED** [review]
I have a question around deleting the set when its empty that I'd like answered before we merge, but tbh i'm still not going to block on it.

not blocking

I'm _currently_ missing why we need to order the sets at all.

I see why here in this implementation, since we are making the tx set the base "object". What I'm wondering is if we need to create that object or if it is possible to add sequence and signer to the existing schema and preserve global priority.

File: mempool/cat/store.go (hunk 10 above)
Comment (cmwaters) [reply mempool/cat/store.go:290]: My intention was to have the store just be about storing transactions and not about reaping them in a certain order which I wanted that logic to be in the pool. I'm just trying to avoid the case where all the logic slides into the store struct and the pool just ends up a wrapper. 

I've seen this style used by databases so I just copied it

File: mempool/cat/tx.go (hunk 11 above)
Comment (cmwaters) [reply mempool/cat/tx.go:87]: Yeah this is a good catch. We should just be doing sequence and not trying to split them by timestamp. I think we can rely on CheckTx to ensure that the case of two transactions with the same signer and sequence don't occur

**Comment by cmwaters** [comment]
> I'm currently missing why we need to order the sets at all.

I'm trying to solve this problem: Imagine we have two transactions A and B with sequences 1 and 2 respectively. If B is higher paying when we go to recheck the transaction it will be processed before A and rejected for having the wrong sequence even although A does exist and as long as they are executed in the right order both can exist in the mempool and in a single block. 

Why the sets of transactions need to be ordered? - because we want a strategy of including transactions that pays the most fees. 

I hope that answers your question - happy to do a 5 min sync to go over it if it doesn't
